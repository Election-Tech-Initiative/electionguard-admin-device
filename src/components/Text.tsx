import React from 'react'
import styled, { css } from 'styled-components'

import GLOBALS from '../config/globals'

interface Props {
  bold?: boolean
  center?: boolean
  error?: boolean
  italic?: boolean
  muted?: boolean
  narrow?: boolean
  small?: boolean
  warning?: boolean
  warningIcon?: boolean
  wordBreak?: boolean
  white?: boolean
  voteIcon?: boolean
}

const iconStyles = css<Props>`
  &::before {
    display: inline-block;
    margin-top: -0.3rem;
    margin-right: 0.25rem;
    border-radius: ${({ warningIcon }) => (warningIcon && '50%') || undefined};
    background: ${({ warningIcon, voteIcon }) =>
      (warningIcon && 'darkorange') || (voteIcon && '#028099')};
    width: 1rem;
    height: 1rem;
    vertical-align: middle;
    text-align: center;
    line-height: 1.1;
    color: #ffffff;
    font-size: 90%;
    font-weight: 800;
    content: ${({ warningIcon, voteIcon }) =>
      (warningIcon && `'!'`) || (voteIcon && `'${GLOBALS.CHECK_ICON}'`)};
  }
`

const Text = styled('p')<Props>`
  margin-right: ${({ narrow }) => (narrow ? 'auto' : undefined)};
  margin-left: ${({ narrow }) => (narrow ? 'auto' : undefined)};
  max-width: ${({ narrow }) => (narrow ? '33ch' : undefined)};
  text-align: ${({ center }) => (center ? 'center' : undefined)};
  color: ${({ error, muted, warning, white }) =>
    (error && 'red') ||
    (warning && 'darkorange') ||
    (white && '#FFFFFF') ||
    (muted && 'gray') ||
    undefined};
  @media print {
    color: ${({ error, muted, warning, white }) =>
      (error && 'black') ||
      (warning && 'black') ||
      (white && '#FFFFFF') ||
      (muted && 'black') ||
      undefined};
  }
  font-size: ${({ small }) => (small ? '0.8rem' : undefined)};
  font-weight: ${({ bold }) => (bold ? '600' : undefined)};
  font-style: ${({ italic }) => (italic ? 'italic' : undefined)};
  word-break: ${({ wordBreak }) => (wordBreak ? 'break-word' : undefined)};
  /* stylelint-disable-next-line value-keyword-case, order/properties-order */
  ${({ warningIcon, voteIcon }) => (warningIcon || voteIcon) && iconStyles}
`

export const TextWithLineBreaks = ({ text }: { text: string }) => (
  <>
    {text.split(/[\n|\r]{2}/g).map(x => (
      <p key={x}>
        {x.split(/[\n|\r]/g).map((y, i, arr) => (
          <React.Fragment key={y}>
            {y}
            {arr.length > 1 && i !== arr.length - 1 && <br />}
          </React.Fragment>
        ))}
      </p>
    ))}
  </>
)

export default Text
