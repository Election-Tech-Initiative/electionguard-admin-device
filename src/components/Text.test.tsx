import React from 'react'
import { render } from '@testing-library/react'

import Text, { TextWithLineBreaks } from './Text'

it(`outputs paragraph tag`, async () => {
  const text = 'paragraph'
  const { getByText } = render(<Text>{text}</Text>)
  const element = getByText(text)
  expect(element.tagName).toEqual('P')
  expect(element).toMatchSnapshot()
})

it(`outputs "span" tag specified by "as" prop`, async () => {
  const text = 'Text in a span?'
  const { getByText } = render(<Text as="span">{text}</Text>)
  const element = getByText(text)
  expect(element.tagName).toEqual('SPAN')
  expect(element).toMatchSnapshot()
})

it(`centers centered style`, async () => {
  const { container } = render(<Text center>Centered Text?</Text>)
  expect(container.firstChild).toMatchSnapshot()
})

it(`displays muted style`, async () => {
  const { container } = render(<Text muted>Muted Text?</Text>)
  expect(container.firstChild).toMatchSnapshot()
})

it(`displays error style`, async () => {
  const { container } = render(<Text error>Error Text?</Text>)
  expect(container.firstChild).toMatchSnapshot()
})

it(`converts line-breaks into <p> and <br/> tags`, () => {
  const { container } = render(
    <TextWithLineBreaks
      text={`I’m a paragraph.\n\nAnd I’m a paragraph with a\nline break.`}
    />
  )
  expect(container).toMatchInlineSnapshot(`
<div>
  <p>
    I’m a paragraph.
  </p>
  <p>
    And I’m a paragraph with a
    <br />
    line break.
  </p>
</div>
`)
})
