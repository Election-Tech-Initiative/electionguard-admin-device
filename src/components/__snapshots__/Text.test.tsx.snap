// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`centers centered style 1`] = `
<p
  class="sc-bdVaJa bApkjD"
>
  Centered Text?
</p>
`;

exports[`displays error style 1`] = `
<p
  class="sc-bdVaJa iGcpUZ"
>
  Error Text?
</p>
`;

exports[`displays muted style 1`] = `
<p
  class="sc-bdVaJa WYgMq"
>
  Muted Text?
</p>
`;

exports[`outputs "span" tag specified by "as" prop 1`] = `
<span
  class="sc-bdVaJa gemIuj"
>
  Text in a span?
</span>
`;

exports[`outputs paragraph tag 1`] = `
<p
  class="sc-bdVaJa gemIuj"
>
  paragraph
</p>
`;
